src/streams.h

```diff
@@ -221,60 +222,25 @@ class CDataStream
     typedef vector_type::reverse_iterator reverse_iterator;
 
     explicit CDataStream(int nTypeIn, int nVersionIn)
-    {
-        Init(nTypeIn, nVersionIn);
-    }
-
-    CDataStream(const_iterator pbegin, const_iterator pend, int nTypeIn, int nVersionIn) : vch(pbegin, pend)
-    {
-        Init(nTypeIn, nVersionIn);
-    }
-
-    CDataStream(const char* pbegin, const char* pend, int nTypeIn, int nVersionIn) : vch(pbegin, pend)
-    {
-        Init(nTypeIn, nVersionIn);
-    }
-
-    CDataStream(const vector_type& vchIn, int nTypeIn, int nVersionIn) : vch(vchIn.begin(), vchIn.end())
-    {
-        Init(nTypeIn, nVersionIn);
-    }
-
-    CDataStream(const std::vector<char>& vchIn, int nTypeIn, int nVersionIn) : vch(vchIn.begin(), vchIn.end())
-    {
-        Init(nTypeIn, nVersionIn);
-    }
+        : nType{nTypeIn},
+          nVersion{nVersionIn} {}
 
-    CDataStream(const std::vector<unsigned char>& vchIn, int nTypeIn, int nVersionIn) : vch(vchIn.begin(), vchIn.end())
-    {
-        Init(nTypeIn, nVersionIn);
-    }
+    explicit CDataStream(Span<const uint8_t> sp, int nTypeIn, int nVersionIn)
+        : vch(sp.data(), sp.data() + sp.size()),
+          nType{nTypeIn},
+          nVersion{nVersionIn} {}
 
     template <typename... Args>
     CDataStream(int nTypeIn, int nVersionIn, Args&&... args)
+        : nType{nTypeIn},
+          nVersion{nVersionIn}
     {
-        Init(nTypeIn, nVersionIn);
         ::SerializeMany(*this, std::forward<Args>(args)...);
     }
 
-    void Init(int nTypeIn, int nVersionIn)
+    void Rewind()
```

Comment:
There is already a `bool Rewind(size_type n)`, overloading it here is confusing. I we should think of a new name. I had hoped to think of a way to use the existing method but it appears there is  no way to access `nReadPos` from the outside.

Reply:
I think the name make sense. `Rewind(size_t)` rewinds only how much is passed. `Rewind()` rewinds all the way. Though, I am happy to rename this to anything else, if there are suggestions.

Reply:
It's a pity that there is no defined invalid value for size_t, which could be passed as default value. Would `Rewind(std::optional<size_t> n = std::nullopt)` be a possibility here to unite the two methods? Like e.g.:

```diff
diff --git a/src/streams.h b/src/streams.h
index 2a921c2c5..a6e781b95 100644
--- a/src/streams.h
+++ b/src/streams.h
@@ -14,6 +14,7 @@
 #include <assert.h>
 #include <ios>
 #include <limits>
+#include <optional>
 #include <stdint.h>
 #include <stdio.h>
 #include <string.h>
@@ -238,11 +239,6 @@ public:
         ::SerializeMany(*this, std::forward<Args>(args)...);
     }

-    void Rewind()
-    {
-        nReadPos = 0;
-    }
-
     std::string str() const
     {
         return (std::string(begin(), end()));
@@ -339,12 +335,17 @@ public:
         nReadPos = 0;
     }

-    bool Rewind(size_type n)
+    bool Rewind(std::optional<size_type> n = std::nullopt)
     {
+        // Total rewind if no size is passed
+        if (!n) {
+            nReadPos = 0;
+            return true;
+        }
         // Rewind by n characters if the buffer hasn't been compacted yet
-        if (n > nReadPos)
+        if (*n > nReadPos)
             return false;
-        nReadPos -= n;
+        nReadPos -= *n;
         return true;
     }
```

Probably overkill for  methods that are only ever called in the benchmarks though...

Reply:
Thanks, fixed

Reply:
> I think the name make sense. Rewind(size_t) rewinds only how much is passed.

Right. My remark was triggered mostly by that the two implementations of `rewind()` were in different places in the method list and not together, the name makes some sense.

But I like the new implementation.